storage/database.go

```diff
@@ -158,34 +172,220 @@ func NewDatabase(
 		return nil, errCommitLogStrategyUnknown
 	}
 
-	ns := make(map[ts.Hash]databaseNamespace, len(namespaces))
-	blockRetrieverMgr := opts.DatabaseBlockRetrieverManager()
+	mediator, err := newMediator(d, opts)
+	if err != nil {
+		return nil, err
+	}
+	d.mediator = mediator
+
+	// initialize namespaces
+	nsInit := opts.NamespaceInitializer()
+	nsReg, err := nsInit.Init()
+	if err != nil {
+		return nil, err
+	}
+
+	// get a namespace watch
+	watch, err := nsReg.Watch()
+	if err != nil {
+		return nil, err
+	}
+
+	// wait till first value is recieved
+	<-watch.C()
+	d.nsWatch = newDatabaseNamespaceWatch(d, watch, iopts)
+
+	// update with initial values
+	nsMap := watch.Get()
+	if err := d.updateOwnedNamespaces(nsMap); err != nil {
+		return nil, err
+	}
+
+	return d, nil
+}
+
+func (d *db) updateOwnedNamespaces(newNamespaces namespace.Map) error {
+	d.Lock()
+	defer d.Unlock()
+
+	removes, adds, updates := d.namespaceDeltaWithLock(newNamespaces)
+	if err := d.logNamespaceUpdate(removes, adds, updates); err != nil {
+		enrichedErr := fmt.Errorf("unable to log namespace updates: %v", err)
```

Comment:
Perhaps a better log message is `unable to update existing namespaces while database open: %v`?